### PR TITLE
Storage sparse indexdb cache

### PR DESCRIPTION
**Pull request by zekker6**
Related issue: https://github.com/VictoriaMetrics/VictoriaMetrics/issues/7182

- add a separate index cache for searches which might read through large amounts of random entries. Primary use-case for this is retention and downsampling filters, when applying filters background merge needs to fetch large amount of random entries which pollutes an index cache. Using different caches allows to reduce effect on memory usage and cache efficiency of the main cache while still having high cache hit rate. A separate cache size is 5% of allowed memory.

- reduce size of indexdb/dataBlocks cache in order to free memory for new sparse cache. Reduced size by 5% and moved this to a separate cache.

- add a separate metricName search which does not cache metric names - this is needed in order to allow disabling metric name caching when applying downsampling/retention filters. Applying filters during background merge accesses random entries, this fills up cache and does not provide an actual improvement due to random access nature.

Merge performance and memory usage stats before and after the change:

- before
image

- after
image

### CODE REVIEW

**Comment by f41gh7** [comment]
Note, currently this change increases performance only for Enteprise version with enabled `downsampling` or `retention` filters.

File: app/vmstorage/main.go [hunk 1]
```diff
@@ -67,6 +67,8 @@ var (
 		"See https://docs.victoriametrics.com/single-server-victoriametrics/#cache-tuning")
 	cacheSizeIndexDBDataBlocks = flagutil.NewBytes("storage.cacheSizeIndexDBDataBlocks", 0, "Overrides max size for indexdb/dataBlocks cache. "+
 		"See https://docs.victoriametrics.com/single-server-victoriametrics/#cache-tuning")
+	cacheSizeIndexDBDataBlocksSparse = flagutil.NewBytes("storage.cacheSizeIndexDBDataBlocksSparse", 0, "Overrides max size for indexdb/dataBlocksSparse cache. "+
```
Comment (hagen1778) [inline app/vmstorage/main.go:70]: If this can be overriden, it should have been mentioned in https://docs.victoriametrics.com/#cache-tuning
I am ok with not having it overridable, though. As it seems too complicated to explain and shoulb be something the system adjusts on its own.

File: lib/storage/index_db.go [hunk 2]
```diff
@@ -520,15 +521,29 @@ type indexSearch struct {
...
 }
 
+// getIndexSearch returns an indexSearch with default configuration
 func (db *indexDB) getIndexSearch(deadline uint64) *indexSearch {
+	return db.getIndexSearchInternal(deadline, false)
+}
+
+// getIndexSearchSparse returns an indexSearch with sparse cache
+// It is useful for search operations that can scan through the large amount index entries
+// Without the need to keep all the entries in the cache
```
Comment (hagen1778) [inline lib/storage/index_db.go:531]:
```suggestion
// Without the need to keep all the entries in the caches used for queries
```

File: docs/changelog/CHANGELOG.md [hunk 3]
```diff
@@ -25,6 +25,7 @@ See also [LTS releases](https://docs.victoriametrics.com/lts-releases/).
...
 * FEATURE: [vmgateway](https://docs.victoriametrics.com/vmgateway/): allow parsing `scope` claim parsing in array format. This is useful for cases when identity provider does encode claims in array format.
 * FEATURE: [vmui](https://docs.victoriametrics.com/#vmui): add the ability to cancel running queries. See [this issue](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/7097).
+* FEATURE: [Single-node VictoriaMetrics](https://docs.victoriametrics.com/) and `vmstorage` in [VictoriaMetrics cluster](https://docs.victoriametrics.com/cluster-victoriametrics/): add a separate cache type for storing sparse entries when performing large index scans. This significantly reduces memory usage when applying [downsampling](https://docs.victoriametrics.com/#downsampling) and [retention filters](https://docs.victoriametrics.com/#retention-filters) during background merge. See [this issue](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/7182) for the details.
```
Comment (hagen1778) [inline docs/changelog/CHANGELOG.md:28]: Does it affect downsampling or downsampling-filters only?

Comment (zekker6) [reply docs/changelog/CHANGELOG.md:28]: Only downsampling filters. I'll rephrase this to make it clear

File: app/vmstorage/main.go (hunk 1 above)
Comment (zekker6) [reply app/vmstorage/main.go:70]: As far as I can see there is explicit mentions for any caches at this page, only generic guidance about cache tuning. Should we mention this one specifically?

Comment (hagen1778) [reply app/vmstorage/main.go:70]: Yeah, my bad. Quickly glanced the page without actually reading it(

**Review by hagen1778 — APPROVED** [review]
LGTM
